spylib/utils/session_token.py

```diff
@@ -0,0 +1,122 @@
+from __future__ import annotations
+
+from typing import Any, ClassVar, Dict, List, Optional
+from urllib.parse import urlparse
+
+from jwt import decode
+from pydantic.main import BaseModel
+from pydantic.networks import HttpUrl
+
+from .domain import store_domain
+
+
+class ValidationError(Exception):
+    pass
+
+
+class InvalidIssuerError(ValidationError):
+    pass
+
+
+class MismatchedHostError(ValidationError):
+    pass
+
+
+class TokenAuthenticationError(ValidationError):
+    pass
+
+
+class SessionToken(BaseModel):
+    """
+    Session tokens are derived from the authrization header from Shopify.
+    This performs the set of validations as defined by shopify
+    https://shopify.dev/apps/auth/session-tokens/authenticate-an-embedded-app-using-session-tokens#obtain-session-details-manually
+    """
+
+    iss: HttpUrl
+    dest: HttpUrl
+    aud: Optional[str]
+    sub: int
+    exp: Optional[float]
+    nbf: Optional[float]
+    iat: Optional[float]
+    jti: int
+    sid: str
+
+    algorithm: ClassVar[str] = "HS256"
+    prefix: ClassVar[str] = "Bearer "
+    required_fields: ClassVar[List[str]] = ["iss", "dest", "sub", "jti", "sid"]
```

Comment:
Do we really want to make that customizable?
You could use the `@property` trick to make them class constants and organized.

Reply:
Good point, will fix.

Reply:
Actually bit of a problem, `@property` isn't itterable. So  `required_fields` is going to be a problem as it will require a custom class to implement the itterable property. 

I also have to cast every call to the `prefix` property back to string as all of the functions are expecting values of type `str` not `property` type.

I will wait and see if you think this is worth implementing or if it is over-complicating it.

Reply:
I am confused ... This works perfectly with `mypy` for me. Maybe I'm not triggering the right problem
```python
class PropertyTest:
    def __init__(self):
        pass

    @property
    def prop(self) -> list:
        return [1, 2, 3, 4]

    @property
    def propstr(self) -> str:
        return 'Success'

protest = PropertyTest()

for i in protest.prop:
    print(i)

print(type(protest.propstr))
```
```bash
1
2
3
4
<class 'str'>
```

Reply:
Sorry should have been more specific, for the type checking on the input of some of the functions I get errors:

`TypeError: startswith first arg must be str or a tuple of str, not property`

I can 'fix' this by casting the property to a string but that won't work as it doesn't actually contain the value. 

I found a fix, properties contain a method `fget()` which you can use to obtain the literal from the method.

Reply:
Just kidding, better fix is to have to set it as a classmethod, which works for the default while a static method and method will not?